### DISCUSSION TITLE
Inlay hint on vim9

### DISCUSSION THREAD

**Discussion by chemzqm** [neoclide/coc.nvim#4095] [Show and tell]
Requires vim >= 9.0.0067, use latest vim >= 9.0.0252 to avoid some serious bugs.

<img width="960" alt="Screen Shot 2022-08-24 at 22 15 47" src="https://user-images.githubusercontent.com/251450/186442060-d1b3efb3-9ed7-40db-a76d-bb8c9cd90703.png">

Configurations needed to enable inlay hint on typescript, like:
```json
  "typescript.inlayHints.variableTypes.enabled": true,
  "typescript.inlayHints.parameterNames.enabled": "all",
  "typescript.inlayHints.enumMemberValues.enabled": true,
  "typescript.inlayHints.parameterTypes.enabled": true,
  "typescript.inlayHints.functionLikeReturnTypes.enabled": true,
  "typescript.inlayHints.propertyDeclarationTypes.enabled": true,
```

Check the doc by `:h coc-inlayHint`

**Comment by 1player** [comment]
Where are these new settings like `typescript.inlayHints.variableTypes.enabled` documented? They're not in the wiki nor in the docs.

I want to disable all type hints in Rust and I'm not sure how to do it.

**Reply by chemzqm** [reply]
https://github.com/neoclide/coc-tsserver#configuration-options or install coc-json for LSP features.

**Comment by oblitum** [comment]
Just learned of this feature, finally, it's great. Now I wish nvim follows asap... and also, even for vim9, there's a lot of coc plugins that need to be updated I guess, I know coc-pyright has been, not sure about coc-rust-analyzer, coc-clangd, etc, they all suffered a lot from the limitation of hints solely at the end of line, it was more confusing noise than helpful in a lot of cases, too difficult to read.

**Reply by fannheyward** [reply]
Extensions don't need to change anything, this has been done in coc side.

**Reply by oblitum** [reply]
That sounds really great then, but then I don't understand why coc-pyright required [this PR](https://github.com/fannheyward/coc-pyright/pull/797).

**Reply by fannheyward** [reply]
Inlay hints are came from language server, but Pyright doesn't support this. coc-pyright added this in extension, that PR just changes hints positions.

**Reply by oblitum** [reply]
But hint position is precisely the relevant aspect I'm referring to regarding possible changes. I know inlay hints by virtual text has been supported long ago for nvim, and now for vim9, but do coc-rust-analyzer and coc-clangd now have it for vim9, but also, those with hints in the middle of text, without any modification, given they all just had it at the end of line on nvim? Just a simple yes or no on that can clears things up for me a bit.

I see clangd and rust-analyzer are a different beast than pyright because they do provide inlay hints from language server, but those could never be displayed as in VSCode in middle text, until now with Vim9 implementation that allows it. There's even a coc-rust-analyzer issue where it was discussed it isn't supported, previously.

**Reply by fannheyward** [reply]
1. inlayHint is now part of LSP, coc.nvim supports this, includes fetching hints from LS, displaying on vim/nvim
2. coc can display hints in the middle of text on vim, but not nvim
3. coc-pyright supports inlayHint in extension: only returns hints to coc, with label and position infos, coc displays them
4. coc-rust-analyzer has removed inlay hints supports from extension because coc handled this
5. coc-clangd supports both because older clangd's inlay hints protocol is different from LSP's, with older clangd, coc-clangd displays hints on nvim only, for newer clangd, coc handles this
6. yes, coc-rust-analzyer and coc-clangd(with newer clangd) have it for vim 9+

In short, inlayHint is now supported in coc with fetching, displaying, toggling etc. vim9 supports inlay hint better than nvim now.

**Reply by oblitum** [reply]
Many thanks for detailed reply. I'm really envy of this vim feature now, but my setup is all lua these days, which I rather prefer... can't decide.

**Reply by fannheyward** [reply]
My opinion: Lua is better on performances, plugin developments, lower FFI API etc, but not for configurations. I'm still using init.vim.
